### DISCUSSION TITLE
Error: on initilizing input_size in DeepAREstimator

### DISCUSSION THREAD

**Discussion by AsthaSatsangi** [zalandoresearch/pytorch-ts#54] [Q&A]
Getting error: `input.size(-1) must be equal to input_size. Expected 43, got 19`
when we run estimator with the dataset showed in the summary text of your post using the example and dateset : 

```py
url = "https://raw.githubusercontent.com/numenta/NAB/master/data/realTweets/Twitter_volume_AMZN.csv"
df = pd.read_csv(url, header=0, index_col=0, parse_dates=True)

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

estimator = DeepAREstimator(freq="5min",
                            prediction_length=12,
                            input_size=43,
                            trainer=Trainer(epochs=10,
                                            device=device))
predictor = estimator.train(training_data=training_data, num_workers=4)
Error:input.size(-1) must be equal to input_size. Expected 43, got 19
```

Please explain how input_size is 43 and how the code is getting it as 19.

**Comment by kashif** [comment]
Thanks for the question... 

So yes in PyTorch we need to explicitly create the correct sized layers and this `input_size` reflects the size of the vector input to the RNN. Now this vector input to the RNN consists of obviously the `target` which is just 1-dim for the univariate case at each time point and then we concat to this the covariates which are typically time features made from the `freq` string... And it's in the covariate features that the input size can vary... For minute data for example we have minute of hour, hour of day day of week, etc. 

Also depending on the data set we can have any number of dynamic features and also embedding vectors coming from categorical features, whose size is another hyper-parameter. 

Therefore the upshot of it is that the `input_size` is a bit difficult to calculate a-priori... On bad solution is to just use the error message to set the `input_size`. A more elegant solution would be to first do all the transformations, feature creation etc. and then instantiate the model... In any case I hope this clears the mystery?

**Reply by cmapz2** [reply]
Thanks!

**Reply by tanagha** [reply]
Can you please elaborate on the Statement:  "A more elegant solution would be to first do all the transformations, feature creation etc" 
I am having a hard time understanding the part of the transformations made in the deepar_estimator.py file which is the source code of the model. 
Thankyou for the help.
